lib/shared/s3-packages.js

```diff
@@ -0,0 +1,245 @@
+ *
+ * @param {RELEASE_TYPE} releaseType - release type
+ * @returns {?String} S3 bucket url
+ *
+ * @example
+ * const bucketUrl = s3Packages.getBucketUrlFromReleaseType(release.RELEASE_TYPE.PRODUCTION);
+ *
+ * if (bucketUrl) {
+ *   console.log(bucketUrl);
+ * }
+ */
+exports.getBucketUrlFromReleaseType = (releaseType) => {
+  if (releaseType === release.RELEASE_TYPE.PRODUCTION) {
+    return exports.BUCKET_URL.PRODUCTION;
+  }
+
+  if (releaseType === release.RELEASE_TYPE.SNAPSHOT) {
+    return exports.BUCKET_URL.SNAPSHOT;
+  }
+
+  return null;
+};
+
+/**
+ * @summary Get all remote versions from an S3 bucket
+ * @function
+ * @private
+ *
+ * @description
+ * We memoize based on the assumption that the received latest version
+ * number will not increase while the application is running.
+ *
+ * @param {String} bucketUrl - s3 bucket url
+ * @fulfil {String[]} - remote versions
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.getRemoteVersions(s3Packages.BUCKET_URL.PRODUCTION).then((versions) => {
+ *   _.each(versions, (version) => {
+ *     console.log(version);
+ *   });
+ * });
+ */
+exports.getRemoteVersions = _.memoize((bucketUrl) => {
+  if (_.isNil(bucketUrl)) {
+    return Bluebird.reject(new Error(`Invalid bucket url: ${bucketUrl}`));
+  }
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return request.getAsync(bucketUrl)
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    .get('body')
+    .then(xml.parseStringAsync)
+    .get('ListBucketResult')
+    .then((bucketResult) => {
+      return _.get(bucketResult, [ 'Contents' ], []);
+    })
+    .reduce((accumulator, entry) => {
+      const [ name, version ] = _.split(_.first(entry.Key), '/');
+
+      if (name === S3_PACKAGE_NAME) {
+        if (_.isNil(accumulator[version])) {
+          accumulator[version] = 1;
+        } else {
+          accumulator[version] += 1;
+        }
+      }
+
+      return accumulator;
+    }, [])
+    .then((versions) => {
+      return _.keys(_.pickBy(versions, (occurrences) => {
+        return occurrences >= NUMBER_OF_PACKAGES;
+      }));
+    })
+    .catch({
+      code: 'ENOTFOUND'
+    }, {
+      code: 'ETIMEDOUT'
+    }, () => {
+      return [];
+    });
+});
+
+/**
+ * @summary Check if a version satisfies a semver range
+ * @function
+ * @private
+ *
+ * @description
+ * This function is a wrapper around `semver.satisfies`
+ * to make it work fine with pre-release versions.
+ *
+ * @param {String} version - semver version
+ * @param {String} range - semver range
+ * @returns {Boolean} whether the version satisfies the range
+ *
+ * @example
+ * if (semverSatisfies('1.0.0', '>=1.0.0')) {
+ *   console.log('The version satisfies the range');
+ * }
+ */
+const semverSatisfies = (version, range) => {
+
+  // The `semver` module refuses to apply ranges to prerelease versions
+  // As a workaround, we drop the prerelease tags, if any, apply the range
+  // on that, and keep using the prerelease tag from then on.
```

Comment:
Perhaps we're getting our wires crossed. That's certainly how we're _using_ it, but the `semverSatisfies` function itself only returns a `Boolean`, and so 'the function itself' has no idea that we "keep using the prerelease tag" after the function has returned. Bit of a subtle point, which is why I added the smiley at the end!